compiler/plugins/target/AMD-AIE/aievec/VectorToVectorConversions.cpp

```diff
@@ -925,6 +927,156 @@ struct CanonicalizeVectorForAIEVecPass
+
+  Value oldIndex = readOp.getIndices().back();
+
+  Value offset = rewriter.createOrFold<affine::AffineApplyOp>(
+      loc, moduloMap, SmallVector<Value, 1>{oldIndex});
+
+  Value elementBytes =
+      rewriter.create<arith::ConstantIndexOp>(loc, elementBits / 8);
+
+  Value offsetBytes =
+      rewriter.createOrFold<arith::MulIOp>(loc, offset, elementBytes);
+
+  Value offsetBytes_i32 = rewriter.createOrFold<arith::IndexCastOp>(
+      loc, rewriter.getIntegerType(32), offsetBytes);
+
+  // If the offset is constant and zero, the read is already aligned.
+  if (auto offsetConstantOp = offset.getDefiningOp<arith::ConstantIndexOp>())
+    if (offsetConstantOp.getValue() == 0) return readOp.getVector();
+
+  // Verify that we can load a vector 2x as long as the original vector.
+  int64_t longBits = 2 * shortBits;
+  int64_t longLength = 2 * shortLength;
+  VectorType longType = VectorType::get(longLength, elementType);
+  if (longBits > maxVectorSizeBits) {
+    // Not returning failure, as it is possible that the read is already
+    // aligned, and we just couldn't prove it.
+    readOp.emitWarning()
+        << "`transfer_read` can't be aligned with a read twice "
+        << "as large because " << longBits
+        << " bits is greater than the maximum vector size of "
+        << maxVectorSizeBits << " bits.";
+
+    return readOp.getVector();
+  }
+
+  Value newIndex = rewriter.createOrFold<arith::SubIOp>(loc, oldIndex, offset);
+
+  SmallVector<bool> inBounds = readOp.getInBoundsValues();
+  bool allInBounds =
+      std::all_of(inBounds.begin(), inBounds.end(), [](bool b) { return b; });
+
+  if (shortBytes != 64 && shortBytes != 32) {
+    // Not returning failure, as it is possible that the read is already
+    // aligned, and we just couldn't prove it.
+    readOp.emitWarning()
+        << "`transfer_read` doesn't have a vector with 64 or 32 bytes. "
+        << "This case is not currently handled.";
+    return readOp.getVector();
+  }
+
+  // Create the aligned transfer read for a vector 2x as long that covers the
+  // elements of the unaligned vector.
+  Value longVec = rewriter.create<vector::TransferReadOp>(
+      loc, longType, readOp.getSource(), SmallVector<Value>{newIndex}, padding,
+      SmallVector<bool>{allInBounds});
+
+  Value replacement = [&]() -> Value {
+    if (shortBytes == 64) {
+      // - Extract lower 64 bytes
+      // - Extract upper 64 bytes
+      // - Apply shift to obtain new 64 bytes
+      Value low = rewriter.create<ExtOp>(loc, shortType, longVec,
+                                         rewriter.getI8IntegerAttr(0));
+      Value upp = rewriter.create<ExtOp>(loc, shortType, longVec,
+                                         rewriter.getI8IntegerAttr(1));
+      return rewriter.createOrFold<ShiftOp>(loc, shortType, low, upp,
+                                            offsetBytes_i32);
+    }
+    assert(shortBytes == 32 && "Expect nb bytes 64 or 32.");
```

Comment:
I think it would be better to throw an error as `shortBytes` depends on data type + shape, so this might actually be hit in a use case from what I can tell?